### ISSUE TITLE
Discussion for IHostingEnvironment updates

### ISSUE THREAD

**Issue by JunTaoLuo** [aspnet/Hosting#651]
Discussion for https://github.com/aspnet/Announcements/issues/157

We made the following changes in our updates in `IHostingEnvironment`:
- Added `ApplicationName`. This replaces `IApplicationEnvironment.ApplicationName` but otherwise contain the same information.
- Added `ContentRootPath`. This replaces `IApplicationEnvironment.ApplicationBasePath` but otherwise contain the same information.
- Removed Configuration
- Removed MapPath

**Comment by halter73** [comment]
@DamianEdwards I think `ApplicationContentBasePath` like you suggested in https://github.com/aspnet/PlatformAbstractions/issues/26 is better than `ContentRootPath`. The latter *might* have made sense if it were still on `IApplicationEnvironment` (even then...), but I really disagree now that it's on `IHostingEnvironment`.

In a vacuum, I'd totally expect `IHostingEnvironment.ContentRootPath` to bet the path for static assets. Only if I carefully looked at all the properties on `IHostingEnvironment` and then read the doc comments would I realize I was wrong.

Other possible alternatives:

- ApplicationContentPath
- AspNetRootPath
- ConfigurationPath

**Comment by JunTaoLuo** [comment]
I like `ApplicationContentPath` slightly better. Or maybe `ApplicationContentRoot` or `ApplicationContentBase` since we know that it is a path.

**Comment by davidfowl** [comment]
`ApplicaitonFilesPath` bhahah

**Comment by DamianEdwards** [comment]
I don't think any of the suggested names are better than what we have already.

**Comment by pranavkm** [comment]
An FYI for folks using `dotnet run` with an Mvc application serving Views - with this change, your applications would have to use `dotnet run -- --contentRoot $PWD` to allow Mvc to locate views correctly.

**Comment by cwe1ss** [comment]
From the announcement:
> This means that the fields on IApplicationEnvironment, specifically ApplicationName and ApplicationBasePath, will no longer be set.

Will `IApplicationEnvironment` be removed? If not, under which circumstances would we still have to use it and what would we have to do if you no longer set these properties?

**Comment by JunTaoLuo** [comment]
@cwe1ss You should be using `IHostingEnvironment` instead as it contains the information we used to set in `IApplicationEnvironment. You should be using `IHostingEnvironment.ApplicationName` and `IHostingEnvironment.ContentRootPath`.

**Comment by shaunluttin** [comment]
What differentiates `ContentRootPath` from `WebRootPath`?

**Comment by GuardRex** [comment]
@shaunluttin `ContentRootPath` is the new name for the application base path. `WebRootPath` is the path to your static assets (`wwwroot` or whatever you decide to call it).

**Comment by shaunluttin** [comment]
@GuardRex Thank you for that. It still instead clear, though, when to use the one or the other. For instance, is this appropriate?

    var contentRoot = Directory.GetCurrentDirectory();
    var webRoot = Path.Combine(contentRoot, "wwwroot");

It doesn't seem to be, because in IIS, we will then have content root as `wwwroot` and webroot as `wwwroot/wwwroot`. So, I am not sure how to handle that. Right now a workaround is like this, but it doesn't seem good. 

    // if content root is `wwwroot`, then move it up one directory
    contentRoot = contentRoot.Replace("wwwroot", string.Empty);

So in IIS, the content root path is the web root path? Help me understand here.

**Comment by GuardRex** [comment]
@shaunluttin If you have a `wwwroot` folder in your project (at the `ContentRootPath` path, the app base path) and you specify `wwwroot` in the `content` section of `project.json`, the `wwwroot` folder will deploy to the `ContentRootPath` ... so it really will just end up being `ContentRootPath` something like `F:\mysite` and `WebRootPath` along with it `F:\mysite\wwwroot`. See the structure for published output here: https://github.com/dotnet/cli/issues/937 (latest is at the bottom of the issue). Therefore ...
```
var contentRoot = Directory.GetCurrentDirectory();
var webRoot = Path.Combine(contentRoot, "wwwroot");
```
... is correct.

**Comment by shaunluttin** [comment]
@GuardRex Ok. So I just published using `dotnet publish --output publish`. Here is the result.

    MyApp
        project.json
        publish                                   <---- the publish output
            MyApp.exe
            wwwroot
                web.config
        wwwroot
            web.config

The I point IIS at `publish/wwwroot`.

The following is in my `Configure()`.

    context.Response.WriteAsync(_hostingEnv.ContentRootPath + "\r\n");
    context.Response.WriteAsync(_hostingEnv.WebRootPath + "\r\n");

And this is in my `Main()`.

    var contentRoot = Directory.GetCurrentDirectory();
    var webRoot = Path.Combine(contentRoot, "wwwroot");

    .UseContentRoot(contentRoot)
    .UseWebRoot(webRoot)

The output on the screen is not what I would expect. 

    C:\GitHub\StackOverflow\AspNetCoreSqlite\publish\wwwroot
    C:\GitHub\StackOverflow\AspNetCoreSqlite\publish\wwwroot\wwwroot

**Comment by GuardRex** [comment]
@shaunluttin Point IIS to the `ContentRootPath` (application base path). Place your `web.config` right out on the app base path with the app's files. `wwwroot` will truly only contain static assets going forward. [Don't forget to add `web.config` to the `content` node of `project.json`.]

The reason you're getting `C:\GitHub\StackOverflow\AspNetCoreSqlite\publish\wwwroot` as the `ContentRootPath` is because when the app gets going from the `wwwroot` folder (where you pointed IIS to your `web.config` file), it thinks that that path is the `ContentRootPath`.

When you move the `web.config` file to the app base path (out of `wwwroot`), IIS will point to the app base path and start on that path and the `ContentRootPath` will be the same. Then, the `WebRootPath` will properly be `C:\GitHub\StackOverflow\AspNetCoreSqlite\publish\wwwroot`.

If you look at the current samples at https://github.com/aspnet/cli-samples, you'll see this type of project layout. Checkout the `HelloMvc` sample for example.

These types of changes are in the Annoucements for RC2: https://github.com/aspnet/Announcements/issues?q=is%3Aopen+is%3Aissue+milestone%3A1.0.0-rc2

**Comment by shaunluttin** [comment]
@GuardRex How does that translation into publishing to Azure? For instance, our publish has the following structure. 

```
publish
  refs
  wwwroot
  MyApp.db
  MyApp.exe
  web.config
```

Do we just dump the entirety into Azure's `/site` directory?

**Comment by GuardRex** [comment]
@shaunluttin Yes, you will just dump the contents of the app (the contents of the `publish` folder) to that directory (`site`). Just waiting on them to roll-out the AspNetCoreModule to Azure Apps. That's in the works right now, but there is no ETA yet on when it will happen. It could be any day now.

**Comment by shaunluttin** [comment]
@GuardRex Does it continue to work with the `httpPlatform` module? I.e. like this: 

<?xml version="1.0" encoding="utf-8"?>

```
<configuration>
  <system.webServer>
    <handlers>
      <add name="httpPlatformHandler"
           path="*"
           verb="*"
           modules="httpPlatformHandler"
           resourceType="Unspecified" />
    </handlers>
    <httpPlatform processPath=".\MyApp.exe"
                  arguments="%DNX_ARGS%"
                  stdoutLogEnabled="false"
                  startupTimeLimit="3600" />
  </system.webServer>
</configuration>
```

**Comment by GuardRex** [comment]
@shaunluttin ... no ... it won't work. There was a period before AspNetCoreModule arrived that you could use RC2 bits with HPH (HttpPlatformHandler), but recent testing by me and another dev show that current RC2-nightly (today's bits) will not work with HPH. Another dev and I tested that scenario on both an Azure VM and Azure Apps ... it failed in both places.

**Comment by shaunluttin** [comment]
@GuardRex That's a bummer. It's odd too, because it works locally with IIS.

**Comment by GuardRex** [comment]
Yes! That's true. It does work locally. The other dev had the same result. I don't think it will take them that long to roll out that module. Milestones look very good for RC2 release, so I don't think it will be long now.

![capture](https://cloud.githubusercontent.com/assets/1622880/14547885/9f90c1b6-0277-11e6-9f4b-20959958c0e6.PNG)

**Comment by shaunluttin** [comment]
@GuardRex One more question... how does Azure know to look for web.config in the `site` directory instead of the `site/wwwroot` directory?

**Comment by GuardRex** [comment]
That config I believe is part of the :sparkles: magic :sparkles: in Azure Apps sauce. They must be setting the path in IIS under-the-covers for Azure Apps. Normally, IIS will scan and consume `web.config` files in the directory structure of an app (sub-folders), so maybe part of their behind-the-scenes updates for the new module will be to re-point IIS under Azure Apps to start looking for `web.config` files one directory higher than it used to ... instead of looking at `site/wwwroot`, perhaps they back that out to starting with `site` folder. I'm **wildly** guessing how the :sparkles: magic :sparkles: works.

**Comment by GuardRex** [comment]
@shaunluttin btw ---- Join us on JabbR if you have ?'s ... better place to talk to a lot of top gun devs. https://jabbr.net/#/rooms/AspNetCore

**Comment by yzorg** [comment]
Would the team consider "WebApplicationPath" instead of "ContentRootPath"?  To me the web application includes what VS calls "content files" (configuration files, project.json), the dotnet binaries in ./bin, and the static web files in ./wwwroot.

I get that with the unification of ASP.NET into the dotnet CLI that ApplicationBasePath now needs to point to the ./bin directory, but why not use "Web" prefix to mean the directory above that?  I don't think anyone agrees with the VS definition of "content".

Is the team intentionally staying away from the word "Web" in property names?

**Comment by tuespetre** [comment]
- `ContentRootPath` is an awful name. Whether it's a 'hosting' environment or not, it's still an **application** that's running, so it should be `ApplicationBasePath`. :moneybag: 
- Using `Directory.GetCurrentDirectory()` to set the `ContentRootPath` seems wrong. 
    - It makes it impossible to run the application from another working directory.
    - There should be something in addition to or similar to `PlatformServices.Default.Application.ApplicationBasePath` that indicates the path from which the running project is "truly" based, i.e. the folder containing the `project.json` if that's the case or the specific folder with the exe in a published app's case.

**Comment by twilliamsgsnetx** [comment]
Trying to implement URL rewriting that Steve Sanderson has created for the SpaServices. When I create a HomeController that I intend to serve my html file with, I inject IHostingEnvironment. When I do that however, I get an internal server error that says

ArgumentException: Type 'Microsoft.AspNetCore.Hosting.IHostingEnvironment' does not have a default constructor

I've not run into this before. Am I injecting it wrong?

        public IActionResult Index(IHostingEnvironment hostingEnvironment)
        {
        }

**Comment by davidfowl** [comment]
> I've not run into this before. Am I injecting it wrong?

Yes.

It should be:

```C#
public IActionResult Index([FromServices]IHostingEnvironment hostingEnvironment)
{
}
```

**Comment by muratg** [comment]
We are closing this issue because no further action is planned for this issue. If you still have any issues or questions, please log a new issue with any additional details that you have.
